README.md
# estoque

A Clojure library designed to ... learn Clojure :)

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-10-19
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-10-19
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/estoque/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/estoque/compare/0.1.0...0.1.1


src/estoque/aula2.clj
(ns estoque.aula2)

(defn imprime-mensagem []
  (println "------------------------")
  (println "Bem vindo(a) ao estoque!"))

(imprime-mensagem)



; * valor-bruto 0.9

(defn aplica-desconto [valor-bruto]
  (* valor-bruto 0.9))


(println (aplica-desconto 100))


(defn valor-descontado
  "Retorna o valor descontado que é 90% do valor bruto"
  [valor-bruto]
  (* valor-bruto 0.9))

(println
  (valor-descontado 100))


(defn valor-descontado [valor-bruto] (* valor-bruto 0.9))


(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (* valor-bruto (- 1 0.10)))

(valor-descontado 100)




(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (def desconto 0.10)
  (* valor-bruto (- 1 desconto)))

(valor-descontado 100)




(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto 0.10]
    (* valor-bruto (- 1 desconto))
    )
  )

(valor-descontado 100)











(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto 0.10]
    (println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))
    )
  )

(valor-descontado 100)









(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (
    let [desconto 0.10]
    (
      println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))
    )
  )

(valor-descontado 100)






(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto 0.10]
    (println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))))

(valor-descontado 100)




(class 90.0)



(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [desconto (/ 10 100)]
    (println "Calculando desconto de " desconto)
    (* valor-bruto (- 1 desconto))))

(valor-descontado 100)



(class 90N)
(class 90M)








(defn valor-descontado
  "Retorna o valor com desconto de 10%."
  [valor-bruto]
  (let [taxa-de-desconto (/ 10 100)
        desconto (* valor-bruto taxa-de-desconto)]
    (println "Calculando desconto de " desconto)
    (- valor-bruto desconto)))

(valor-descontado 100)





(> 500 100)
(< 500 100)

(if (> 500 100)
  (println "maior")
  (println "menor ou igual"))

(if (> 50 100)
  (println "maior")
  (println "menor ou igual"))









(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (> valor-bruto 100)
    (let [taxa-de-desconto (/ 10 100)
          desconto (* valor-bruto taxa-de-desconto)]
      (println "Calculando desconto de " desconto)
      (- valor-bruto desconto))))

(valor-descontado 1000)

(valor-descontado 1000)






(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (> valor-bruto 100)
    (let [taxa-de-desconto (/ 10 100)
          desconto (* valor-bruto taxa-de-desconto)]
      (println "Calculando desconto de " desconto)
      (- valor-bruto desconto))
    valor-bruto))

(valor-descontado 1000)

(valor-descontado 100)



src/estoque/aula4.clj
(ns estoque.aula4)

(def precos [30 70 100])

(println (precos 0))
(println (get precos 0))
(println (get precos 2))
(println (get precos 17))
;(println (precos 17))
(println (get precos 17 "XX"))
(println (get precos 2 "XX"))

(println (conj precos 5))
(println precos)

(println (+ 5 1))
(println (inc 4))

(println (update precos 0 inc))
(println precos)

(defn soma+1
  [valor]
  (println "estou somando um a" valor)
  (+ valor 1))

(println (update precos 0 soma+1))
(println precos)


;AULA ANTERIOR

(defn aplica-desconto?
  [valor-bruto]
  (> valor-bruto 100))

(defn valor-descontado
  "Retorna o valor com desconto de 10% se o valor bruto for estritamente maior que 100."
  [valor-bruto]
  (if (aplica-desconto? valor-bruto)
    (let [taxa-de-desconto (/ 10 100)
          desconto (* valor-bruto taxa-de-desconto)]
      (- valor-bruto desconto))
    valor-bruto))

(def precos [30 700 1000])
(println precos)
(println "MAP" (map valor-descontado precos))

(println (range 10))
(println (filter even? (range 10)))                       ;even é se é par
(println "filter" (filter aplica-desconto? precos))

;transformar uma sequencia em outra, ex somar o primeiro valor com o segundo valor
; pq nao usar map ou filter, porque eles só recebem um a um
(println (reduce + precos))

(defn minha-soma
  [valor-1 valor-2]
  (println "somando" valor-1 valor-2)
  (+ valor-1 valor-2))

(println (reduce minha-soma precos))
(println (reduce minha-soma (range 10)))
(println (reduce minha-soma [15]))
(println (reduce minha-soma 0 precos))






src/estoque/aula5.clj
(ns estoque.aula5)

(def estoque {"Mochila" 10,
              "Camiseta" 5})

(println estoque)

(println "temos" (count estoque) "elementos")
(println "Chave" (keys estoque))                            ;nao existe garantia na ordem
(println "Valores" (vals estoque))

(def estoque {:mochila 10,
              :camiseta 5})
(println (assoc estoque :cadeira 3))
(println (assoc estoque :mochila 1))
(println (update estoque :mochila inc))
(println (update estoque :mochila #(- % 3)))
(println (dissoc estoque :mochila))

;-----

(def pedido {:mochila {:quantidade 2, :preco 80}
             :camiseta {:quantidade 3, :preco 40}})

(def pedido (assoc pedido :chaveiro {:quantidade 1, :preco 10}))

(println pedido)
(println (get pedido :mochila))
(println (get pedido :cadeira 0))
(println (:mochila pedido))                                 ;:mochila invoca ifn
(println (:cadeira pedido {}))                              ;mais comum

(println (:quantidade (:mochila pedido)))
(println (update-in pedido [:mochila :quantidade] inc))


(println (:quantidade (:mochila pedido)))

;THREADING (ENCADEANDO, "fiozinho de costura") "tredin" FORMA MAIS USADA DO Q OUTRA
;THREADING FIRST ,,,
(println (-> pedido
             :mochila
             :quantidade))

;pedido.get (mochila).get (quantidade)

; nao tao bom de usar
(-> pedido
    :mochila
    :quantidade
    println)

(def clientes { :15 { :nome "Guilherme" :certificados ["Clojure" "Java" "Machine Learning"] } })
(-> clientes :15 :certificados count)



src/estoque/aula6.clj
(ns estoque.aula6)

(def pedido {:mochila  {:quantidade 2, :preco 80}
             :camiseta {:quantidade 3, :preco 40}})

(defn imprime-e-15 [valor]
  (println "valor" valor (class valor))
  15)

(println (imprime-e-15 24))

(println (map imprime-e-15 pedido))

;destruct (desistruturando o map em chave e valor)
(defn imprime-e-15 [[chave valor]]
  (println chave "<e>" valor)
  valor)

(println (map imprime-e-15 pedido))

(defn preco-dos-produtos [[_ valor]]                        ;_ quer dizer "nao vou usar"
  (* (:quantidade valor) (:preco valor)))

(println (map preco-dos-produtos pedido))

(println (reduce + (map preco-dos-produtos pedido)))

(defn total-do-pedido
  [pedido]
  (reduce + (map preco-dos-produtos pedido)))

(println (total-do-pedido pedido))

;agora com threading LAST: ,,, é a colecao em ultimo
(defn total-do-pedido
  [pedido]
  (->> pedido
      (map preco-dos-produtos ,,,)
      (reduce + ,,,)))

(println (total-do-pedido pedido))

;MELHOR ABORDAGEM
(defn preco-total-do-produto [produto]
  (* (:quantidade produto) (:preco produto)))

(defn preco-total-do-produto
  [pedido]
  (->> pedido
       vals
       (map preco-dos-produtos ,,,)
       (reduce + ,,,)))

(println (total-do-pedido pedido))

;----


(def pedido {:mochila  {:quantidade 2, :preco 80}
             :camiseta {:quantidade 3, :preco 40}
             :chaveiro {:quantidade 1}})

(defn gratuito?
  [item]
  ;(println item)
    (<= (get item :preco 0) 0))

(println "filtrando gratuitos")
(println (filter gratuito? (vals pedido)))
;---


[(defn gratuito?
  [[_ item]]
  ;(println item)
  (<= (get item :preco 0) 0))]

  (println "filtrando gratuitos")
  (println (filter gratuito? pedido))

;---
[(defn gratuito?
   [item]
   (println item)
   (<= (get item :preco 0) 0))]

(println (filter #(gratuito? (second %)) pedido))

;---

(defn pago?
  [item]
  (not (gratuito? item)))

(println (pago? {:preco 50}))
(println (pago? {:preco 0}))

;---

(def pago? (comp not gratuito?))
(println (pago? {:preco 50}))
(println (pago? {:preco 0}))

;---

(def clientes [{ :nome "Guilherme"
                :certificados ["Clojure" "Java" "Machine Learning"] }
               { :nome "Paulo"
                :certificados ["Java" "Ciência da Computação"] }
               { :nome "Daniela"
                :certificados ["Arquitetura" "Gastronomia"] }])

;Extraimos os certificados, contamos e depois somamos.
(->> clientes (map :certificados) (map count) (reduce +))


